stellargraph/core/element_data.py

```diff
@@ -284,8 +285,52 @@ def feature_info(self):
         }
 
 
-def _numpyise(d, dtype):
-    return {k: np.array(v, dtype=dtype) for k, v in d.items()}
+class FlatAdjacencyList:
+    """
+    Stores an adjacency list in one contiguous numpy array in a format similar
+    to a CSR matrix. The main difference between a CSR matrix and this data structure
+    is that this class does not store the edge weight, is only provides an `edge_ilocs`
+    which can be used to look up all other edge data.
+
+    Ths class contains:
+
+        - `self.flat`: a numpy array of edge_ilocs grouped by node
+        - `self.splits`: a numpy array used to index `self.flat`
+
+    A node's corresponding edge_ilocs are access by
+
+        ```self.flat[self.splits[node_iloc-1]: self.splits[node_iloc]]```
+
+    Where:
+        - `self.splits[node_iloc - 1]` is where the edge_ilocs corresponding to the node start
+        - `self.splits[node_iloc]` is where the edge_ilocs corresponding to the node end
+
+    """
+
+    def __init__(self, flat_array, splits):
+        self.splits = splits
+        self.flat = flat_array
+
+    def __getitem__(self, idx):
+        if idx == 0:
+            start = 0
+            stop = self.splits[0]
+        elif idx < 0:
+            raise KeyError("Invalid node ilocs")
```

Comment:
`GraphSAGE` etc return -1 if no node is found. A method is attempting to access this with `-1`  (like I was accidently doing and failing tests) is almost certainly a bug and will return neighbours when there shouldnt' be any.

Reply:
Can we instead fix GraphSAGE (and HinSAGE)?

Reply:
I guess this can be follow-up work.

Reply:
hmm... this -1 behaviour is also used in `node_features` and is tested in quite a few difference places so I'd prefer to not change it here if at all.

Reply:
Hm, I guess I'm confused. From:

> GraphSAGE etc return -1 if no node is found. A method is attempting to access this with -1 (like I was accidently doing and failing tests) is almost certainly a bug and will return neighbours when there shouldnt' be any.

It sounds like you think we shouldn't support passing in `-1` (in that doing so is a bug), while from

>  See bellow, but unfortunately, our graphsage methods attempt to lookup -1 node ilocs if an isolated head node is passed!

It sounds like GraphSAGE does pass in `-1`?

Given all that, I agree we should reject `-1` as a node iloc for neighbour queries... but it sounds like we might need to update GraphSAGE to not do any neighbour queries on isolated nodes? (e.g. have its own `if iloc == -1` check). As it is, I think this `__getitem__` method never gets called directly, and this branch will never be taken (since everything goes via `get`).

Reply:
oh i see how you're confused, what i mean/think is:

- passing `-1` to `__getitem__` should be a bug 
- passing `-1` to `get` should be allowed and should return the default value

If that makes sense? This should also be consistent with how the previous dictionary based adj lists perform.

Reply:
That distinction makes sense, but I'm not sure it's something we should continue with.

I think this low level code should be as controlled and inflexible as possible, because that generally means speed. In particular, I think we should regard it as an error to pass in an invalid node iloc or ID into `neighbors`, and it's basically accidental that we weren't handling this properly before.

The intention with the use of `get` with the dictionaries was just to handle isolated nodes, since those didn't appear in the dictionaries, whereas every node appears in these adjacency lists.

In any case, GraphSAGE passing in -1 was a bug from #1267 where a check for `None` wasn't updated to `-1`:

https://github.com/stellargraph/stellargraph/blob/9b81deb26191f226a1a213a1c304a3fbe03c45fc/stellargraph/data/explorer.py#L679-L681

I think you fixed it in #1253 (current develop):

https://github.com/stellargraph/stellargraph/blob/ef81955da232007bd7d2c731308515dd783aeac3/stellargraph/data/explorer.py#L681-L685

Reply:
ah yep you're right, I thought we were still trying to look up -1s i'll get rid of `.get` and the negative index check in `.__getitem__`

Reply:
> the negative index check in .__getitem__

I think it's fine to keep this as a check for now, given we're always working with a single node ID (if we were doing bulk queries it'd be a bit of a harder choice). It's useful to catch things similar to the problems in GraphSAGE, especially while we're still transitioning between approaches.